### PR TITLE
Fixes issue #769 - Implement DefaultWebApplication.getResourcePaths

### LINKED ISSUE #769: Implement DefaultWebApplication.getResourcePaths

### PR DESCRIPTION
**Pull request by Thihup**
Maybe we need to revisit this issue to validate better how it will work with Piranha Micro, because it uses the Shrinkwrap protocol.
Fixes #769

### CODE REVIEW

File: webapp/impl/src/test/java/cloud/piranha/webapp/impl/DefaultWebApplicationTest.java [hunk 1]
```diff
@@ -511,6 +512,31 @@ public void testGetResourcePaths() {
         assertNull(webApp.getResourcePaths("/this_will_be_null/"));
     }
 
+    @Test
+    public void testGetResourcePaths2() {
+        DefaultResourceManager resourceManager = new DefaultResourceManager();
+        DefaultWebApplication webApp = new DefaultWebApplication();
+        webApp.setResourceManager(resourceManager);
+        webApp.addResource(new DirectoryResource("."));
```
Comment (manorrock) [inline webapp/impl/src/test/java/cloud/piranha/webapp/impl/DefaultWebApplicationTest.java:520]: Please create an exploded webapp in src/test/webapp to use for testing. E.g src/test/webapp/resourcepaths.

File: webapp/impl/src/main/java/cloud/piranha/webapp/impl/DefaultWebApplication.java [hunk 2]
```diff
@@ -1087,6 +1092,16 @@ public InputStream getResourceAsStream(String location) {
         return resourceManager.getResourceAsStream(location);
     }
 
+    private Stream<String> getResourcePathsImpl(String path) throws IOException, URISyntaxException {
+        URL resource = resourceManager.getResource(path);
+        if (resource == null)
+            return Stream.empty();
+        return Files.list(Paths.get(resource.toURI())).map(x -> {
```
Comment (manorrock) [inline webapp/impl/src/main/java/cloud/piranha/webapp/impl/DefaultWebApplication.java:1099]: Indeed as you alluded to, you cannot use the Files / Path APIs here as the URI might or might not be file based

File: webapp/impl/src/main/java/cloud/piranha/webapp/impl/DefaultWebApplication.java [hunk 3]
```diff
@@ -1087,6 +1092,16 @@ public InputStream getResourceAsStream(String location) {
         return resourceManager.getResourceAsStream(location);
     }
 
+    private Stream<String> getResourcePathsImpl(String path) throws IOException, URISyntaxException {
+        URL resource = resourceManager.getResource(path);
```
Comment (manorrock) [inline webapp/impl/src/main/java/cloud/piranha/webapp/impl/DefaultWebApplication.java:1096]: The ResourceManager API exposes a getAllLocations stream that gives you the entire list of resources. You probably want to use it and then filter what actually applies in this case.

File: webapp/impl/src/test/java/cloud/piranha/webapp/impl/DefaultWebApplicationTest.java [hunk 4]
```diff
@@ -511,6 +512,31 @@ public void testGetResourcePaths() {
...
+        assertNotNull(resourcePathsCatalog);
+        assertTrue(resourcePathsCatalog.contains("/src/main/"));
+        assertTrue(resourcePathsCatalog.contains("/src/test/"));
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void testGetResourcePaths3() {
+        DefaultWebApplication webApp = new DefaultWebApplication();
+        assertNull(webApp.getResourcePaths(null));
+        webApp.getResourcePaths("");
```
Comment (manorrock) [inline webapp/impl/src/test/java/cloud/piranha/webapp/impl/DefaultWebApplicationTest.java:537]: It looks like you are testing 2 things here. Maybe 2 different tests?

File: webapp/impl/src/main/java/cloud/piranha/webapp/impl/DefaultWebApplication.java [hunk 5]
```diff
@@ -1087,6 +1089,25 @@ public InputStream getResourceAsStream(String location) {
...
+        String normalizedPath = path.endsWith("/") ? path : path + "/";
+        String[] split = resource.replace(normalizedPath, "/").split("/");
+        // It's a directory
+        if (split.length > 2)
+            return normalizedPath + split[1] + "/";
+        // It's a file
+        return normalizedPath + split[1];
+    }
+
+    private Set<String> getResourcePathsImpl(String path) {
```
Comment (manorrock) [inline webapp/impl/src/main/java/cloud/piranha/webapp/impl/DefaultWebApplication.java:1102]: Please add JavaDoc explaining what this does

File: webapp/impl/src/main/java/cloud/piranha/webapp/impl/DefaultWebApplication.java [hunk 6]
```diff
@@ -1087,6 +1089,25 @@ public InputStream getResourceAsStream(String location) {
         return resourceManager.getResourceAsStream(location);
     }
 
+    private String getFileOrFirstFolder(String path, String resource){
```
Comment (manorrock) [inline webapp/impl/src/main/java/cloud/piranha/webapp/impl/DefaultWebApplication.java:1092]: Please add JavaDoc explaining what this does

**Comment by Thihup** [comment]
It's not the scope of this pull request, but I could see that there are two resource managers, and the method getAllLocations returns the resources of only one. One is declared in the DefaultResourceClassLoader and the other is the DefaultResourceManager. This way, it is not possible to get the resources of jars that the DefaultResourceClassLoader contains. I will certainly need some help in how to proceed.

**Comment by manorrock** [comment]
That is also by design. One is used for web resources (which is what you are tackling right now) and the other is used for class loading. They must be kept separate and each will need to be setup properly with respect to the contract they are implementing. If we are missing something in either one of them it needs to be addressed. Thanks for looking into this!

**Pull request merged**